test/common/router/router_test.cc

```diff
@@ -807,6 +812,150 @@ TEST_F(RouterTest, EnvoyUpstreamServiceTime) {
+  EXPECT_TRUE(verifyHostUpstreamStats(1, 0));
+}
+
+TEST_F(RouterTest, AppendCluster0) { testAppendCluster(absl::nullopt); }
+
+TEST_F(RouterTest, AppendCluster1) { testAppendCluster(absl::make_optional("x-custom-cluster")); }
+
+void RouterTestBase::testAppendUpstreamHost(absl::optional<std::string> hostname_header_name,
+                                            absl::optional<std::string> host_address_header_name) {
+  auto debug_config = std::make_unique<DebugConfig>();
+  debug_config->append_upstream_host_ = true;
+  if (hostname_header_name.has_value()) {
+    debug_config->hostname_header_ = *hostname_header_name;
+  }
+  if (host_address_header_name.has_value()) {
+    debug_config->host_address_header_ = *host_address_header_name;
+  }
+  callbacks_.streamInfo().filterState().setData(DebugConfig::key(), std::move(debug_config),
+                                                StreamInfo::FilterState::StateType::ReadOnly);
+  cm_.conn_pool_.host_->hostname_ = "scooby.doo";
+
+  NiceMock<Http::MockStreamEncoder> encoder;
+  Http::StreamDecoder* response_decoder = nullptr;
+  EXPECT_CALL(cm_.conn_pool_, newStream(_, _))
+      .WillOnce(Invoke([&](Http::StreamDecoder& decoder, Http::ConnectionPool::Callbacks& callbacks)
+                           -> Http::ConnectionPool::Cancellable* {
+        response_decoder = &decoder;
+        callbacks.onPoolReady(encoder, cm_.conn_pool_.host_);
+        return nullptr;
+      }));
+  expectResponseTimerCreate();
+
+  Http::TestHeaderMapImpl headers;
+  HttpTestUtility::addDefaultHeaders(headers);
+  router_.decodeHeaders(headers, true);
+
+  Http::HeaderMapPtr response_headers(new Http::TestHeaderMapImpl{{":status", "200"}});
+  EXPECT_CALL(cm_.conn_pool_.host_->outlier_detector_, putHttpResponseCode(200));
+  EXPECT_CALL(callbacks_, encodeHeaders_(_, true))
+      .WillOnce(Invoke([&hostname_header_name, &host_address_header_name](Http::HeaderMap& headers,
+                                                                          bool) {
+        const Http::HeaderEntry* hostname_header = headers.get(
+            Http::LowerCaseString(hostname_header_name.has_value() ? *hostname_header_name
+                                                                   : "x-envoy-upstream-hostname"));
+        EXPECT_NE(nullptr, hostname_header);
+        EXPECT_EQ("scooby.doo", hostname_header->value().getStringView());
+
+        const Http::HeaderEntry* host_address_header = headers.get(Http::LowerCaseString(
+            host_address_header_name.has_value() ? *host_address_header_name
+                                                 : "x-envoy-upstream-host-address"));
+        EXPECT_NE(nullptr, host_address_header);
+        EXPECT_EQ("10.0.0.5:9211", host_address_header->value().getStringView());
+      }));
+  response_decoder->decodeHeaders(std::move(response_headers), true);
+  EXPECT_TRUE(verifyHostUpstreamStats(1, 0));
+}
+
+TEST_F(RouterTest, AddDefaultHostResponseHeaders00) {
```

Comment:
Can you add one liner `//` above each of these tests explaining in natural language what they do?

Reply:
Done, although I guess one could take issue with how "natural" my language is. Let me know if it looks ok :)